File: leetcode/src/01657_determine_if_two_strings_are_close/solution.go
package leetcode

import "sort"

func closeStrings(word1 string, word2 string) bool {
	freq1 := counter(word1)
	freq2 := counter(word2)
	set1 := counterKeys(freq1)
	set2 := counterKeys(freq2)
	cnt1 := counterValues(freq1)
	cnt2 := counterValues(freq2)
	return setRuneEqual(set1, set2) && sliceIntEqual(cnt1, cnt2)
}

func counter(word string) map[rune]int {
	freq := map[rune]int{}
	for _, ch := range word {
		freq[ch]++
	}
	return freq
}

func counterKeys(freq map[rune]int) map[rune]struct{} {
	res := make(map[rune]struct{}, len(freq))
	for ch := range freq {
		res[ch] = struct{}{}
	}
	return res
}

func counterValues(freq map[rune]int) []int {
	res := make([]int, 0, len(freq))
	for _, v := range freq {
		res = append(res, v)
	}
	sort.Ints(res)
	return res
}

func sliceIntEqual(s1 []int, s2 []int) bool {
	if len(s1) != len(s2) {
		return false
	}
	for i := range s1 {
		if s1[i] != s2[i] {
			return false
		}
	}
	return true
}

func setRuneEqual(s1 map[rune]struct{}, s2 map[rune]struct{}) bool {
	if len(s1) != len(s2) {
		return false
	}
	for k := range s1 {
		if _, ok := s2[k]; !ok {
			return false
		}
	}
	return true
}

File: leetcode/src/01657_determine_if_two_strings_are_close/solution.rs
use std::collections::HashMap;
use std::collections::HashSet;

struct Solution;

impl Solution {
    pub fn close_strings(word1: String, word2: String) -> bool {
        let counter = |w: &str| {
            w.chars()
                .fold(HashMap::<char, usize>::new(), |mut map, ch| {
                    *map.entry(ch).or_default() += 1;
                    map
                })
        };
        let freq1 = counter(&word1);
        let freq2 = counter(&word2);
        let mut cnt1 = freq1.values().collect::<Vec<_>>();
        cnt1.sort_unstable();
        let mut cnt2 = freq2.values().collect::<Vec<_>>();
        cnt2.sort_unstable();
        let set1 = freq1.keys().collect::<HashSet<_>>();
        let set2 = freq2.keys().collect::<HashSet<_>>();
        cnt1 == cnt2 && set1 == set2
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        let word1 = String::from("abc");
        let word2 = String::from("bca");
        assert!(Solution::close_strings(word1, word2));

        let word1 = String::from("a");
        let word2 = String::from("aa");
        assert!(!Solution::close_strings(word1, word2));

        let word1 = String::from("cabbba");
        let word2 = String::from("abbccc");
        assert!(Solution::close_strings(word1, word2));
    }
}
